### PR TITLE
StateFlow Blog

### PR DESCRIPTION
**Pull request by kevin666aa**
## Why are these changes needed?
Introduce Stateflow, a task-solving paradigm that conceptualizes complex task-solving processes backed by LLMs as state machines.
Introduce how to use GroupChat customized speaker selection function to implement the idea.

Note: Depends on #1791.
New changes are: `2024-02-29-StateFlow/index.mdx` and `agentchat_groupchat_stateflow.ipynb`

## Checks

- [ ] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

File: website/blog/2024-02-29-StateFlow/index.mdx [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        else:
+            # Otherwise, let the scientist to speak
+            return scientist
+
+    elif last_speaker is scientist:
+        # if the scientist speaks but doesn't terminate the task, we return to a build-in 'auto' selection method that
+        # asks the groupchat manager to select the next speaker.
+        return "auto"
+
+groupchat = autogen.GroupChat(
```
Comment (LittleLittleCloud) [inline website/blog/2024-02-29-StateFlow/index.mdx:82]: The usage of this example seems unrelated to this `Stateflow` PR? (#1816)?

Comment (kevin666aa) [reply website/blog/2024-02-29-StateFlow/index.mdx:82]: Hello @LittleLittleCloud, sorry for the confusion. #1816 is attempting to introduce a completely new concept and still needs discussion. Currently, it is best to accomplish the idea with groupchat, with customized functions from #1791.

File: test/agentchat/test_groupchat.py [hunk 2]
```diff
@@ -1,7 +1,7 @@
 #!/usr/bin/env python3 -m pytest
 
 from typing import Any, Dict, List, Optional, Type
-from autogen import AgentNameConflict, Agent, GroupChat
+from autogen import AgentNameConflict, Agent, GroupChat, Agent, GroupChat
```
Comment (qingyun-wu) [inline test/agentchat/test_groupchat.py:4]: Agent, GroupChat are repeating

**Comment by qingyun-wu** [comment]
@kevin666aa could you update the notebook added in this PR and the notebook you added in the other PR (https://github.com/microsoft/autogen/blob/main/notebook/agentchat_groupchat_customized.ipynb) following the new guide here: https://github.com/microsoft/autogen/blob/main/notebook/contributing.md#how-to-get-a-notebook-displayed-on-the-website, so that they can be displayed on the website. Thank you!  

In addition to the guide, you can find examples in Jack's PR: https://github.com/microsoft/autogen/pull/1890

**Comment by qingyun-wu** [comment]
Could you fix the code formatting and doc formatting errors? Thank you!

File: website/blog/2024-02-29-StateFlow/index.mdx [hunk 3]
```diff
@@ -135,7 +135,7 @@ When returning `None`, the group chat will terminate.
 
 
 ## For Further Reading
-* [StateFlow notebook](https://microsoft.github.io/autogen/docs/notebooks/agentchat_groupchat_stateflow/)
-* [GroupChat with Customized Speaker Selection notebook](https://microsoft.github.io/autogen/docs/notebooks/agentchat_groupchat_customized/)
+* [StateFlow notebook](https://github.com/microsoft/autogen/blob/main/notebook/agentchat_groupchat_stateflow.ipynb)
+* [GroupChat with Customized Speaker Selection notebook](https://github.com/microsoft/autogen/blob/main/notebook/agentchat_groupchat_customized.ipynb)
```
Comment (sonichi) [inline website/blog/2024-02-29-StateFlow/index.mdx:139]: Isn't it better to use the web-rendered notebook?
Also, we can use paths like (/docs/...)

**Comment by sonichi** [comment]
@jackgerrits do you know how to fix this error? https://github.com/microsoft/autogen/actions/runs/8217918136/job/22474143866?pr=1826#step:9:65

File: test/test_notebook.py [hunk 4]
```diff
@@ -119,6 +119,21 @@ def test_agentchat_cost_token_tracking(save=False):
...
+)
+def test_agentchat_groupchat_customized(save=False):
+    run_notebook("agentchat_groupchat_customized.ipynb", save=save)
+
+@pytest.mark.skipif(
+    skip or not sys.version.startswith("3.10"),
+    reason="do not run if openai is not installed or py!=3.10",
+)
+def test_agentchat_groupchat_stateflow(save=False):
+    run_notebook("agentchat_groupchat_stateflow.ipynb", save=save)
```
Comment (sonichi) [inline test/test_notebook.py:134]: Adding both to py3.10 will make the test workload imbalanced.

Comment (kevin666aa) [reply test/test_notebook.py]: Got it!

**Comment by kevin666aa** [comment]
> @jackgerrits do you know how to fix this error? https://github.com/microsoft/autogen/actions/runs/8217918136/job/22474143866?pr=1826#step:9:65

I am still looking into the error, any help is appreciated! The notebook renders fine, but the link is broken. I can compile the blog and open the notebook locally.

File: website/blog/2024-02-29-StateFlow/index.mdx [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+def state_transition(last_speaker, groupchat):
+   messages = groupchat.messages
+
+
+   if last_speaker is initializer:
+       # init -> retrieve
+       return coder
+   elif last_speaker is coder:
+       # retrieve: action 1 -> action 2
+       return executor
```
Comment (sonichi) [inline website/blog/2024-02-29-StateFlow/index.mdx:110]: Note: For these two and the "Scientist" case, we can use the built-in constraints. Only the executor case needs a special handler.
This doesn't mean the code must be modified, but something to think about or add a note.

File: website/blog/2024-02-29-StateFlow/index.mdx [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+We recommend implementing the transition logic for each speaker in the customized function. In analogy to a state machine, a state transition function determines the next state based on the current state and input.
+Instead of returning an `Agent` class representing the next speaker, we can also return a string from `['auto', 'manual', 'random', 'round_robin']` to select a default method to use.
+For example, we can always default to the built-in `auto` method to employ a LLM-based group chat manager to select the next speaker.
+When returning `None`, the group chat will terminate.
+
+
+## For Further Reading
+* [StateFlow notebook](https://microsoft.github.io/autogen/docs/notebooks/agentchat_groupchat_stateflow/)
+* [GroupChat with Customized Speaker Selection notebook](https://microsoft.github.io/autogen/docs/notebooks/agentchat_groupchat_customized/)
+* [FSM Group Chat](/blog/2024/02/11/FSM-GroupChat/)
```
Comment (sonichi) [inline website/blog/2024-02-29-StateFlow/index.mdx:140]: The new blog doesn't discuss relations with FSM group chat, which could confuse readers. I'd suggest adding discussions.

File: website/blog/2024-02-29-StateFlow/index.mdx [hunk 7]
```diff
@@ -131,7 +132,7 @@ groupchat = autogen.GroupChat(
 We recommend implementing the transition logic for each speaker in the customized function. In analogy to a state machine, a state transition function determines the next state based on the current state and input.
 Instead of returning an `Agent` class representing the next speaker, we can also return a string from `['auto', 'manual', 'random', 'round_robin']` to select a default method to use.
 For example, we can always default to the built-in `auto` method to employ a LLM-based group chat manager to select the next speaker.
-When returning `None`, the group chat will terminate.
+When returning `None`, the group chat will terminate. Note that some of the transitions, such as "initializer" -> "coder" can be defined with the transition graph
```
Comment (sonichi) [inline website/blog/2024-02-29-StateFlow/index.mdx:135]:
```suggestion
When returning `None`, the group chat will terminate. Note that some of the transitions, such as "initializer" -> "coder" can be defined with the transition graph.
```

File: website/blog/2024-02-29-StateFlow/index.mdx [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+Our evaluation demonstrates the advantages of **StateFlow** and **SF_Agent** over existing methods in terms of both effectiveness and efficiency.
+For example, in the Bash task with GPT-4, **SF_Agent** improves the success rate by 8% and has 4x less cost compared to ReAct prompting.
+
+
+## Implement StateFlow With GroupChat
+We illustrate how to build **StateFlow** with GroupChat. Previous blog [FSM Group Chat](/blog/2024/02/11/FSM-GroupChat/)
+introduces a new feature of GroupChat that allows us to input a transition graph to constrain agent transitions.
+It requires us to use natural language to describe the transition conditions of the FSM in the agent's `description` parameter, and then use an LLM to take in the description and make decisions of the next agent.
+In this blog, we take advantages of a customized speaker selection function passed to the `speaker_selection_method` of the `GroupChat` object.
+This function allows us to custom the transition logic between agents, and can be used together with the transition graph introduced in FSM Group Chat.
```
Comment (joshkyh) [inline website/blog/2024-02-29-StateFlow/index.mdx:58]:
```suggestion
This function allows us to custom the transition logic between agents, and can be used together with the transition graph introduced in FSM Group Chat. The current StateFlow implementation allows the user to override the transition graph.
```

File: website/blog/2024-02-29-StateFlow/index.mdx [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+
+
+```python
+# Define the agents, the code is for illustration purposes and is not executable.
+initializer = autogen.UserProxyAgent(
+   name="Init"
+)
+coder = autogen.AssistantAgent(
+   name="Coder",
+   system_message="""You are the Coder. Write Python Code to retrieve papers from arxiv.."""
```
Comment (joshkyh) [inline website/blog/2024-02-29-StateFlow/index.mdx:82]:
```suggestion
   system_message="""You are the Coder. Write Python Code to retrieve papers from arxiv."""
```

File: website/blog/2024-02-29-StateFlow/index.mdx [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+
+In **StateFlow**, we construct a state machine to control a single LLM, sending different instructions to it at different states.
+We also provide an agent view of our framework, **SF_Agent**, that can use different LLM agents to perform actions at different states.
+In this case, we don't need to add instructions to the context history and call the LLM. Instead, we construct individual agents with pre-set instructions and (potentially) different LLMs.
+AutoGen is the perfect platform to implement **SF_Agent**.
+
+## Experiments
+We evaluate **StateFlow** on the SQL task and Bash task from the InterCode benchmark, with both GPT-3.5-Turbo and GPT-4.
+We construct different **StateFlow** models for each task (See the figure below). But note that most of the states are shared between the two tasks.
+Within each state, we define a sequence of actions to be performed. The most common action sequence is P->M->E, meaning we first send a prompt, then call the LLM to generate a response, and finally execute commands from the response.
```
Comment (joshkyh) [inline website/blog/2024-02-29-StateFlow/index.mdx:34]: I think this is the first mention of P, M, E. Who/what are they?

**Review by joshkyh — COMMENTED** [review]
Looking forward to your paper, @kevin666aa! :)

**Comment by sonichi** [comment]
Could you fix the test error?

**Comment by kevin666aa** [comment]
> Could you fix the test error?

The groupchat_customized notebook requires human input, so I removed it from notebook testing.

**Review by sonichi — APPROVED** [review]

**Comment by jackgerrits** [comment]
@kevin666aa sorry for missing the message. Looks like you fixed it up. 

The relevant log of the error was:
```
  Exhaustive list of all broken links found:
  - Broken link on source page path = /autogen/docs/notebooks/agentchat_groupchat_customized:
     -> linking to /autogen/docs/llm_configuration
  - Broken link on source page path = /autogen/docs/notebooks/agentchat_groupchat_stateflow:
     -> linking to /autogen/docs/llm_configuration
```

That page had moved to `/docs/topics/llm_configuration` but it looks like it's all sorted now
